File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/StringAttributeInputCallback.java
/*
 * Copyright (c) 2019 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth.callback;

import androidx.annotation.Keep;

import org.json.JSONException;
import org.json.JSONObject;

import lombok.Getter;
import lombok.NoArgsConstructor;

/**
 * Implements a Callback for collection of a single identity object attribute from a user.
 */
@Getter
public class StringAttributeInputCallback extends AttributeInputCallback {

    /**
     * The attribute Value
     */
    private String value;

    @Keep
    public StringAttributeInputCallback() {
    }

    /**
     * Constructor for this Callback.
     */
    @Keep
    public StringAttributeInputCallback(JSONObject jsonObject, int index) throws JSONException {
        super(jsonObject, index);
    }

    @Override
    protected void setAttribute(String name, Object value) {
        super.setAttribute(name, value);
        if (VALUE.equals(name)) {
            this.value = (String) value;
        }
    }

    public void setValue(String value) {
        super.setValue(value);
    }

    @Override
    public String getType() {
        return "StringAttributeInputCallback";
    }
}

File: e2e/app/src/main/java/com/example/app/callback/StringAttributeInputCallback.kt
/*
 * Copyright (c) 2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.example.app.callback

import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import org.forgerock.android.auth.callback.StringAttributeInputCallback

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun StringAttributeInputCallback(callback: StringAttributeInputCallback) {

    var input by remember {
        mutableStateOf(callback.value)
    }

    Row(modifier = Modifier
        .padding(4.dp)
        .fillMaxWidth()) {
        OutlinedTextField(
            modifier = Modifier,
            value = input,
            onValueChange = { value ->
                input = value
                callback.value = input
            },
            isError = callback.hasError(),
            supportingText = if (callback.hasError()) {
                @Composable {
                    Text(
                        text = callback.error(),
                        style = MaterialTheme.typography.bodySmall
                    )
                }
            } else null,
            label = { Text(callback.prompt) },
        )
    }
}